File: backend/app/routers/caregiver_profiles.py
from bson import ObjectId
from fastapi import APIRouter, Depends, HTTPException

from ..core.database import get_db
from ..core.patient_resolver import resolve_patient_id

router = APIRouter(prefix="/api/caregiver-profiles", tags=["caregiver-profiles"])


@router.get("")
async def list_caregiver_profiles(patient_id: str = Depends(resolve_patient_id)):
    """List all caregivers linked to the current patient."""
    db = get_db()
    patient = await db["patients"].find_one({"_id": ObjectId(patient_id)})
    if not patient:
        raise HTTPException(status_code=404, detail="Patient not found")

    caregiver_ids = patient.get("caregiver_ids", [])
    if not caregiver_ids:
        return []

    caregivers = []
    for cid in caregiver_ids:
        user = await db["users"].find_one({"_id": ObjectId(cid)})
        if user:
            caregivers.append({
                "id": str(user["_id"]),
                "name": user["name"],
                "email": user["email"],
            })

    return caregivers

File: src/server-routes/caregiverProfiles.ts
import { Router } from "express";
import { ObjectId } from "mongodb";
import { getDb } from "../server-core/database";
import { authMiddleware } from "../server-core/security";
import { resolvePatientId } from "../server-core/patientResolver";

const router = Router();

// GET /api/caregiver-profiles
router.get("/", authMiddleware, resolvePatientId, async (req, res) => {
  try {
    const db = getDb();
    const patient = await db.collection("patients").findOne({ _id: new ObjectId(req.patientId!) });
    if (!patient) {
      res.status(404).json({ detail: "Patient not found" });
      return;
    }

    const caregiverIds: string[] = patient.caregiver_ids || [];
    if (caregiverIds.length === 0) {
      res.json([]);
      return;
    }

    const objectIds = caregiverIds.map((cid) => new ObjectId(cid));
    const users = await db.collection("users").find({ _id: { $in: objectIds } }).toArray();
    const caregivers = users.map((user) => ({
      id: String(user._id),
      name: user.name,
      email: user.email,
    }));

    res.json(caregivers);
  } catch (err) {
    console.error("list caregiver profiles error:", err);
    res.status(500).json({ detail: "Internal server error" });
  }
});

export default router;
